instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py

```diff
@@ -161,6 +177,125 @@ def test_fastapi_excluded_urls_not_env(self):
+        }
+        expected_requests_count_attributes = {
+            "http.method": "GET",
+            "http.host": "testserver",
+            "http.scheme": "http",
+            "http.flavor": "1.1",
+            "http.server_name": "testserver",
+        }
+        metrics_list = self.memory_metrics_reader.get_metrics_data()
+        for metric in (
+            metrics_list.resource_metrics[0].scope_metrics[0].metrics
+        ):
+            for point in list(metric.data.data_points):
+                if isinstance(point, HistogramDataPoint):
+                    self.assertDictEqual(
+                        expected_duration_attributes,
+                        dict(point.attributes),
+                    )
+                    self.assertEqual(point.count, 1)
+                    self.assertAlmostEqual(duration, point.sum, delta=20)
+                if isinstance(point, NumberDataPoint):
+                    self.assertDictEqual(
+                        expected_requests_count_attributes,
+                        dict(point.attributes),
+                    )
+                    self.assertEqual(point.value, 0)
+
+    def test_basic_post_request_metric_success(self):
+        start = default_timer()
+        self._client.post("/foobar")
+        duration = max(round((default_timer() - start) * 1000), 0)
+        metrics_list = self.memory_metrics_reader.get_metrics_data()
+        for metric in (
+            metrics_list.resource_metrics[0].scope_metrics[0].metrics
+        ):
+            for point in list(metric.data.data_points):
+                if isinstance(point, HistogramDataPoint):
+                    self.assertEqual(point.count, 1)
+                    self.assertAlmostEqual(duration, point.sum, delta=30)
+                if isinstance(point, NumberDataPoint):
+                    self.assertEqual(point.value, 0)
+
+    def test_metric_uninstruemnt_app(self):
+        self._client.get("/foobar")
+        self._instrumentor.uninstrument_app(self._app)
+        self._client.get("/foobar")
+        metrics_list = self.memory_metrics_reader.get_metrics_data()
+        for metric in (
+            metrics_list.resource_metrics[0].scope_metrics[0].metrics
+        ):
+            for point in list(metric.data.data_points):
+                if isinstance(point, HistogramDataPoint):
+                    self.assertEqual(point.count, 1)
+                if isinstance(point, NumberDataPoint):
+                    self.assertEqual(point.value, 0)
+
+    def test_metric_uninstrument(self):
+        # instrumenting class and creating app to send request
+        self._instrumentor.instrument()
+        app = self._create_fastapi_app()
+        client = TestClient(app)
+        client.get("/foobar")
+        # uninstrumenting class and creating the app again
+        self._instrumentor.uninstrument()
+        app = self._create_fastapi_app()
+        client = TestClient(app)
+        client.get("/foobar")
```

Comment:
You shouldn't create a new app. You should use the same app instance from above.

Reply:
We can not use the same app instance as the earlier app object was created by patched fastapi library. Once we uninstrument, original fastapi class is restored, and we need to create a new fastapi app object to see the changes. Correct me if I'm wrong.

Reply:
No, the test is to ensure the patching is undone correctly when uninstrumented and the same app instance doesn't produce any more telemetry.

Reply:
Is that behaviour/expectation of `uninstrument` documented anywhere?

Poking around, most of the web API instrumentation libraries I looked at don't do this, e.g. Flask, aiohttp-client, Starlette, FastAPI (i.e. this one) all have their `uninstrument` (and `instrument`) APIs only affect _new_ instances, not existing instances.

Some _do_ unhook existing hooks, e.g., Tornado. I'll note that the Tornado instrumentation can _only_ globally instrument/uninstrument, there isn't an API like `instrument_app` as there is for the ASGI and WSGI code, because it relies on patching the underlying class, rather than hooking `__init__` to insert a middleware in the request flow.

Reply:
> the test is to ensure the patching is undone correctly when uninstrumented and the same app instance doesn't produce any more telemetry.

`uninstrument_app` will uninstrument the existing app and I've added a test for the same but `uninstrument` will just restore the class and all the existing app instances that are already created will not be affected by the change and it'll keep on sending the telemetry data. I don't think we can write test for what you are suggesting.

Reply:
I actually meant specifically for this repo. My list of comparisons was other instrumentation packages here. It might be worth documenting the generic instrumentation behaviours expected in this repo for all instruments, as what should have been a near-trivial PR (add metrics to the existing tracing hooks, because the instrumentation hook setup and ASGI intercept was already done) has turned into a significant refactoring and if that's going to be the case for all the Python instrumentation packages, then it might be worth doing that refactoring _first_ across the repo for the instrumentation packages that need it.

That will also help ensure that any new instrumentation contributions behave correctly from first PR, rather than needing months of reworking _away_ from the existing behaviours. It would also possibly allow some more-generic tests (or at least test-templates) for things like the instrument/uninstrument behaviours being iterated on here.

Reply:
We may have missed this for some instrumentation earlier but we try to ensure that `uninstrument` tests are added and work as expected. For instance, the last two PRs adding metrics for [django](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1208#discussion_r943744431) and [urllib3](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1198/commits/fcb2d42e121f24b68a0ebb4d67ce68abf35e1568) didn't initially have them but later added in review. And same applies to the new instrumentation that gets added (most recent one https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1082) The general guideline help for new instrumentation specifically for this repo is written here https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md#guideline-for-instrumentations. Changes across the repo require some proposal/discussion and followed by implementation. Feel free to volunteer to take that up and help us.

Reply:
Can we merge this PR for now and address the changes in other PR so that it do not get blocked for long?